code/espurna/tuya.cpp

```diff
@@ -27,41 +28,143 @@ Copyright (C) 2019 by Maxim Prokhorov <prokhorov dot max at outlook dot com>
+    constexpr uint32_t HeartbeatFast { 3000u };
+
+    struct Config {
+        Config(const Config&) = delete;
+        Config(Config&& other) noexcept :
+            key(std::move(other.key)),
+            value(std::move(other.value))
+        {}
+
+        Config(String&& key_, String&& value_) noexcept :
+            key(std::move(key_)),
+            value(std::move(value_))
+        {}
+
+        String key;
+        String value;
+    };
+
+    Transport tuyaSerial(TUYA_SERIAL);
+    std::priority_queue<DataFrame> outputFrames;
+
+    template <typename T>
+    void send(unsigned char dp, T value) {
+        outputFrames.emplace(
+            Command::SetDP, DataProtocol<T>(dp, value).serialize()
+        );
+    }
+
+    // --------------------------------------------
+
+    Discovery discovery(DiscoveryTimeout);
+    bool transportDebug { false };
+    bool configDone { false };
+    bool reportWiFi { false };
+    bool filter { false };
+
+    String product;
+    std::forward_list<Config> config;
+
+    DpMap switchIds;
+
+#if LIGHT_PROVIDER == LIGHT_PROVIDER_CUSTOM
+    DpMap channelIds;
+    uint8_t channelStateId { 0u };
+#endif
 
     // --------------------------------------------
 
-    struct dp_states_filter_t {
-        using type = unsigned char;
-        static const type NONE = 0;
-        static const type BOOL = 1 << 0;
-        static const type INT  = 1 << 1;
-        static const type ALL = (INT | BOOL);
+    class TuyaRelayProvider : public RelayProviderBase {
+    public:
+        explicit TuyaRelayProvider(unsigned char dp) :
+            _dp(dp)
+        {}
 
-        static type clamp(type value) {
-            return constrain(value, NONE, ALL);
+        const char* id() const {
+            return "tuya";
         }
+
+        void change(bool status) {
+            send(_dp, status);
+        }
+    private:
+        unsigned char _dp;
+    };
+
+#if LIGHT_PROVIDER == LIGHT_PROVIDER_CUSTOM
+
+    class TuyaLightProvider : public LightProvider {
+    public:
+        TuyaLightProvider() = default;
+        explicit TuyaLightProvider(const DpMap& channels) :
+            _channels(channels)
+        {}
+
+        explicit TuyaLightProvider(const DpMap& channels, uint8_t state) :
+            _channels(channels),
+            _state(state)
+        {}
+
+        void update() override {
+        }
+
+        void state(bool status) override {
+            if (_state) {
+                send(_state, status);
+            }
+        }
+
+        void channel(unsigned char channel, double value) override {
+            auto dp = _channels.dp(channel);
+            if (!dp) {
+                return;
+            }
+
+            // tuya dimmer is precious, and can't handle 0...some-kind-of-threshold
+            // just ignore it, the associated switch will handle turning it off
+            auto rounded = static_cast<unsigned int>(value);
+            if (rounded <= 0x10) {
+                return;
+            }
```

Comment:
Basically, it simply skips any value less than 16 (aka 0x10). This is based on the return value from the dimmer which seemed to stop right after this threshold.